include/quda_matrix.h

```diff
@@ -1068,7 +1069,30 @@ namespace quda {
         - q(3) * q(1) * q(8) - q(0) * q(7) * q(5);
 
       Complex sg2h3 = sqrt(a3 * a3 - (Float)4. * a2 * a2 * a2);
-      Complex cp = exp(log((Float)0.5 * (a3 + sg2h3)) / (Float)3.0);
+
+      // If the matrix q is zero, this algorithm produces nan, instead of unity.
+      // The first invalid operation is the log hereafter. Therefore, we check
+      // if its argument (tmp) is zero and then return unity.
+
+      Complex tmp = a3 + sg2h3;
+      if (tmp.x == 0 and tmp.y==0) {
```

Comment:
Can you switch these `.x` to `.real()` and `.y` to `.imag()` here.  Eventually I'd like to have `std::complex` be a compatible drop in replacement for `quda::complex` (I know there's lots more instances of `.x` and `.y` spread all the place).